KAFKA-14412: Add statestore.uncommitted.max.bytes

nicktelford opened the pull request:
This configuration controls the maximum amount of local state store data to buffer in transaction buffers.

When the total uncommitted data, across all stores and all threads, exceeds this value, a Task commit will be triggered, even if `commit.interval.ms` has not yet elapsed. This prevents application from hitting OutOfMemory errors when the `commit.interval.ms` is set to a high value, and transactional state stores are in-use.

StateStores that support transactions can implement `approximateNumUncommittedBytes` to track the size of the data not yet committed to the store.

By default, `approximateUncommitted` returns `0`, which means that all data is immediately committed to the state store.

nicktelford committed:
KAFKA-14412: Add statestore.uncommitted.max.bytes

This configuration controls the maximum amount of local state store data to
buffer in transaction buffers.

When the total uncommitted data, across all stores and all threads,
exceeds this value, a Task commit will be triggered, even if
`commit.interval.ms` has not yet elapsed. This prevents application from
hitting OutOfMemory errors when the `commit.interval.ms` is set to a
high value, and transactional state stores are in-use.

StateStores that support transactions can implement
`approximateNumUncommittedBytes` to track the size of the data not yet
committed to the store.

By default, `approximateUncommitted` returns `0`, which means that all
data is immediately committed to the state store.

nicktelford commented:
Note: this is part of KIP-892 Transactional StateStores. It introduces a new config option that will have no effect until the rest of KIP-892 lands at a later date.

@cadonna @ableegoldman

nicktelford commented on streams/src/main/java/org/apache/kafka/streams/processor/StateStore.java:
```diff
@@ -165,4 +165,22 @@ default Position getPosition() {
+    /**
+     * Return an approximate count of memory used by records not yet committed to this StateStore.
+     * <p>
+     * This method will return an approximation of the memory that would be freed by the next call to {@link
+     * #flush()}.
+     * <p>
+     * If no records have been written to this store since {@link #init(StateStoreContext, StateStore) opening}, or
+     * since the last {@link #flush()}; or if this store does not support atomic transactions, it will return {@code
+     * 0}, as no records are currently being buffered.
+     *
+     * @return The approximate size of all records awaiting {@link #flush()}; or {@code 0} if this store does not
+     *         support transactions, or has not been written to since {@link #init(StateStoreContext, StateStore)} or
+     *         last {@link #flush()}.
```
Note: the `flush()` references here are a bit misleading. It will all make sense when the main KIP-892 implementation lands and this is renamed to `commit(Map)` though.

nicktelford committed:
Add missing implementations

These various StateStores wrap an inner store, so should delegatee
`approximateNumUncommittedBytes` to that store instad of using the
default value of `0`.

nicktelford committed:
Fix premature commit when buffers are full

The calculation to predict when the transaction buffers would exceed
their capacity was incorrect, and we were failing to do this prediction
at all for some commits.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -505,6 +505,11 @@ public class StreamsConfig extends AbstractConfig {
+    public static final String STATESTORE_UNCOMMITTED_MAX_BYTES_CONFIG = "statestore.uncommitted.max.bytes";
+    public static final String STATESTORE_UNCOMMITTED_MAX_BYTES_DOC = "Maximum number of memory bytes to be used to buffer uncommitted state-store records. " +
+            "If this limit is exceeded, a task commit will be requested. No limit: -1. " +
+            "Note: if this is too high or unbounded, it's possible for RocksDB to trigger out-of-memory errors.";
```
I would not explicitly mention RocksDB here since this config is applicable to any state store that supports transactions.

nicktelford replied:
I've removed that line entirely now.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -874,6 +879,12 @@ public class StreamsConfig extends AbstractConfig {
                     atLeast(0),
                     Importance.MEDIUM,
                     STATESTORE_CACHE_MAX_BYTES_DOC)
+            .define(STATESTORE_UNCOMMITTED_MAX_BYTES_CONFIG,
+                    Type.LONG,
+                    64 * 1024 * 1024L,
+                    atLeast(-1),
+                    Importance.MEDIUM,
+                    STATESTORE_UNCOMMITTED_MAX_BYTES_DOC)
```
Could you please add unit tests for this config in `StreamsConfigTest`?

nicktelford replied:
Done. Let me know if you think I need more tests around this or if the ones I've added are enough.

cadonna replied:
Added a comment to `StreamsConfigTest`.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:
```diff
@@ -438,4 +438,11 @@ public Map<TopicPartition, Long> changelogOffsets() {
         return storeToChangelogTopic.get(storeName);
     }
+
+    @Override
+    public long approximateNumUncommittedBytes() {
+        return globalStores.values().stream()
+            .map(optional -> optional.map(StateStore::approximateNumUncommittedBytes).orElse(0L))
+            .reduce(0L, Long::sum);
+    }
```
I cannot find a unit test for this.

nicktelford replied:
Done.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -754,4 +754,11 @@ public void deleteCheckPointFileIfEOSEnabled() throws IOException {
     }
+
+    @Override
+    public long approximateNumUncommittedBytes() {
+        return stores.values().stream()
+                .map(metadata -> metadata.store().approximateNumUncommittedBytes())
+                .reduce(0L, Long::sum);
+    }
```
I cannot find a unit test for this.

nicktelford replied:
Done.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1374,7 +1376,7 @@ public void signalResume() {
     int maybeCommit() {
         final int committed;
-        if (now - lastCommitMs > commitTimeMs) {
+        if (taskManager.needsCommit(true) || now - lastCommitMs > commitTimeMs) {
```
I could also not find a test code for this.

nicktelford replied:
I've added a test for this in 0577905, but it feels like a bit of a tautology: it essentially just tests that `commit()` is called when `taskManager.needsCommit(anyBoolean())` returns `true`.

The real meat of this will be in the `TaskManagerTests`, which I'm working on separately.

cadonna replied:
We need both tests. The one in `StreamThreadTest` verifies that the commit is triggered or not depending on the value returned from `needsCommit()` and the one in `TaskManagerTest` verifies that `needCommit()` returns the expected boolean value.

Apropos `TaskManagerTest`, I missing a unit test for `needsCommit()` there. Good that you are already working on it.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java:
```diff
@@ -141,6 +141,7 @@ private int processTask(final Task task, final int maxNumRecords, final long beg
     int commitTasksAndMaybeUpdateCommittableOffsets(final Collection<Task> tasksToCommit,
                                                     final Map<Task, Map<TopicPartition, OffsetAndMetadata>> consumedOffsetsAndMetadata) {
         int committed = 0;
+        final boolean enfoceCheckpoint = taskManager.needsCommit(false);
```
typo:
```suggestion
        final boolean enforceCheckpoint = taskManager.needsCommit(false);
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java:
```diff
@@ -157,7 +158,8 @@ int commitTasksAndMaybeUpdateCommittableOffsets(final Collection<Task> tasksToCo
                 ++committed;
-                task.postCommit(false);
+                // under EOS, we need to enforce a checkpoint if our transaction buffers will exceeded their capacity
+                task.postCommit(enfoceCheckpoint);
```
I could not find tests for this change.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
What do you think about moving the `STATESTORE_UNCOMMITTED_MAX_BYTES_CONFIG` to the `InternalConfig` and not exposing this config in the public API for now? Then, when we have a complete implementation, we can expose it publicly. 
This would avoid exposing unfinished work.

nicktelford replied:
Yeah, I can do that. I'm a little concerned that we might forget to move it back though!

cadonna replied:
There are ways to remember. You could tie a knot in your handkerchief or you could create a sub-ticket under  KAFKA-14412. I leave it to you what option you use. 🙂

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1836,6 +1841,33 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
     }
 
+    // track the size of the transaction buffer on each iteration to predict when it will be exceeded in advance
+    private long lastUncommittedBytes = 0L;
```
Please move this field to the other fields at the top of the class.

nicktelford replied:
Done

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1836,6 +1841,33 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
+    private long lastUncommittedBytes = 0L;
+
+    boolean needsCommit(final boolean updateDelta) {
+        if (maxUncommittedStateBytes < 0) {
+            // if our transaction buffers are unbounded, we never need to force an early commit
+            return false;
+        }
```
nit:
```suggestion
        final boolean transactionBuffersAreUnbounded = maxUncommittedStateBytes < 0;
        if (transactionBuffersAreUnbounded) {
            return false;
        }
```

nicktelford replied:
Done

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1836,6 +1841,33 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
+
+        final long deltaBytes = Math.max(0, uncommittedBytes - lastUncommittedBytes);
+
+        final boolean needsCommit =  maxUncommittedStateBytes > -1 && uncommittedBytes + deltaBytes > maxUncommittedStateBytes;
```
Do we really need `maxUncommittedStateBytes > -1`? Isn't this always true once we pass line 1851?

nicktelford replied:
Woops! I did some refactoring and didn't spot the dead code :smile:

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1836,6 +1841,33 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
+
+        final long deltaBytes = Math.max(0, uncommittedBytes - lastUncommittedBytes);
+
+        final boolean needsCommit =  maxUncommittedStateBytes > -1 && uncommittedBytes + deltaBytes > maxUncommittedStateBytes;
```
Additionally, nit:
```suggestion
        final boolean needsCommit = maxUncommittedStateBytes > -1 && uncommittedBytes + deltaBytes > maxUncommittedStateBytes;
```

nicktelford replied:
Done

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java:
```diff
@@ -157,7 +158,8 @@ int commitTasksAndMaybeUpdateCommittableOffsets(final Collection<Task> tasksToCo
                 task.clearTaskTimeout();
                 ++committed;
-                task.postCommit(false);
+                // under EOS, we need to enforce a checkpoint if our transaction buffers will exceeded their capacity
```
Shouldn't our transaction buffers be empty at this point?
We committed the transaction which means we should have also committed the state store.

nicktelford replied:
Actually, at this point we've only committed our Kafka transaction. The StateStore has not been committed/flushed, so our transaction buffers will still be full.

`task.postCommit` is the process that "checkpoints" state stores. We need to enforce a checkpoint when the transaction buffer is full, because we only flush the StateStore when it is checkpointed. See `AbstractTask` line `99`.

cadonna replied:
Ah, OK! Got it!

cadonna replied:
I am also missing unit tests for this case.

nicktelford replied:
There aren't actually any existing tests at all for `TaskExecutor#commitTasksAndMaybeUpdateCommittableOffsets`. I can still add some tests specifically for these changes, but I suspect that a lot more of the behaviour here is relying on integration tests.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1097,10 +1102,10 @@ void handleRevocation(final Collection<TopicPartition> revokedPartitions) {
+                        // we only enforce a checkpoint if the transaction buffers are full
+                        // to avoid unnecessary flushing of stores under EOS
+                        final boolean enforceCheckpoint = maxUncommittedStateBytes > -1 && tasks.approximateUncommittedStateBytes() >= maxUncommittedStateBytes;
+                        task.postCommit(enforceCheckpoint);
```
Also here, the tasks should have been committed. Why are transaction buffers not empty?

nicktelford replied:
Same as above.

That said, I'm not convinced this particular change is necessary. This code is run during rebalance, and AFAIK no records will be processing during a rebalance; it shouldn't be possible for the transaction buffer to have increased beyond the last time we checked in the `StreamThread` main run-loop (aka. `StreamThread#maybeCommit`).

Worst case scenario, if we leave this as `task.postCommit(false)`, and the state stores aren't flushed during a rebalance, if the transaction buffer is already close to exceeding its capacity, it will be detected during the next iteration of the run-loop.

cadonna reviewed (commented):
Thanks for the PR @nicktelford !

I really like how you broke out this PR from the huge PR you had before.

I did a first pass. I had some questions.

Unit tests are missing.

nicktelford committed:
Simplify configuration docs

nicktelford committed:
Add StreamsConfigTests for new config

nicktelford committed:
Add GlobalStateManager tests

nicktelford committed:
Add ProcessorStateManagerTests

nicktelford committed:
Add StreamThreadTest for early commit

Tests that early commits are forced if the TaskManager requests it.

nicktelford committed:
Fix TaskExecutor typo

nicktelford committed:
Tidy up TaskManager#needsCommit

nicktelford committed:
Revert forcing checkpoint during task revocation

There should never be a need to force a checkpoint on task revocation,
because even if the transaction buffers are full (or will be filled
during the next iteration), we will check again during the next
iteration of the StreamThread run-loop, and do the commit+checkpoint at
that point.

nicktelford committed:
Improve readability of TaskManager#needsCommit

nicktelford committed:
Remove unnecessary space from TaskManager

cadonna commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -1320,6 +1320,19 @@ public void shouldUseDefaultStateStoreCacheMaxBytesConfigWhenNoConfigIsSet() {
+        final StreamsConfig config = new StreamsConfig(props);
+        assertEquals(Long.valueOf(64 * 1024 * 1024), config.getLong(StreamsConfig.STATESTORE_UNCOMMITTED_MAX_BYTES_CONFIG));
+    }
+
```
I think you need at least two more tests here. One that tests that you can set positive values and one that verifies that an exception is thrown when you set a value less than -1.

I am wondering if we should allow 0. At the moment, it does not make sense to me to set this config to 0. Maybe you should also write a test with the expected behavior for 0. In the end unit tests should also document the code.

nicktelford replied:
With a value of `0`, it would trigger an "early" commit whenever transaction buffers contain any data. If we decide to disallow this, then we need to determine where we draw the line, because a value of `1` or `2`, for example, would likely yield the same result.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1343,6 +1348,17 @@ private long getCacheSizePerThread(final int numStreamThreads) {
         return totalCacheSize / (numStreamThreads + (topologyMetadata.hasGlobalTopology() ? 1 : 0));
     }
 
+    private long getMaxUncommittedBytesPerThread(final int numStreamThreads) {
```
This method does not consider whether `default.state.isolation.level` is set to `READ_COMMITTED`. Of course before we do not have `default.state.isolation.level`, we cannot consider it. 
Does passing `maxUncommittedBytesPerThread` already affect commit behavior? As far as I understand the code it does not because `approximateNumUncommittedBytes()` returns always 0 at the moment.
Would it be better to first introduce `default.state.isolation.level` before we go on with this?

nicktelford replied:
> Does passing maxUncommittedBytesPerThread already affect commit behavior? As far as I understand the code it does not because approximateNumUncommittedBytes() returns always 0 at the moment.

No, for exactly this reason.

> This method does not consider whether default.state.isolation.level is set to READ_COMMITTED.

Does it need to? Under READ_UNCOMMITTED, since transaction buffers aren't used, `approximateNumUncommittedBytes()` will always return `0` (as it does right now), ensuring that we never exceed the configured threshold.

Notably, if we make this dependent on the isolation level, any future changes that add support for transaction buffers to READ_UNCOMMITTED writes would break this.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1836,6 +1842,30 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
+            log.debug(
+                    "Needs commit because we will exceed max uncommitted bytes before next commit. max: {}, last: {}, current: {}, delta: {}",
+                    maxUncommittedStateBytes, lastUncommittedBytes, uncommittedBytes, deltaBytes
+            );
```
nit:
```suggestion
            log.debug(
                "Needs commit because we will exceed max uncommitted bytes before next commit. max: {}, last: {}, current: {}, delta: {}",
                maxUncommittedStateBytes, lastUncommittedBytes, uncommittedBytes, deltaBytes
            );
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:
```diff
@@ -404,4 +404,11 @@ public synchronized Map<TaskId, Task> allTasksPerId() {
     }
+
+    @Override
+    public long approximateUncommittedStateBytes() {
```
I could not find unit tests for this.

nicktelford replied:
Done

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractSegments.java:
```diff
@@ -152,6 +153,13 @@ public List<S> allSegments(final boolean forward) {
     }
 
+    @Override
+    public long approximateNumUncommittedBytes() {
```
I could not find unit tests for this.

nicktelford replied:
I've added some tests, but since it's not possible to mock the segments themselves, and by default they just return `0` uncommitted bytes, it's not really possible to test that it's correctly summing this until it's possible to enable transaction buffers in the future. I've added a note that the test will consequently need updating later.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1073,6 +1078,35 @@ int commit(final Collection<Task> tasksToCommit) {
+        thread.setNow(mockTime.milliseconds());
+        thread.maybeCommit();
+
+        verify(taskManager, times(2)).commit(mkSet(runningTask));
```
I was wondering why the verification expects 2 invocations. I then understood that when `needsCommit()` returns `false` the commit is done because the second condition of the `if`-statement is `true` (i.e. it is the first commit). The second commit happens because `needsCommit()` returns `true`, but the second condition is `false`. 
This setup is not ideal because it does not test whether `needsCommit()` avoids an early commit when it returns `false`.  So you need a third call to `maybeCommit()`. The first call would just make the first commit. The second call is within the commit interval (i.e., the second condition is `false`) and `needsCommit()` returns `false` and the third call is within the commit interval and `needsCommit()` returns `true`. You should also verify the call to `taskManager.commit()` after each call to `thread.maybeCommit()` to be sure that the call was triggered or not for the correct case.

nicktelford replied:
I've added the third case, and I've independently verified the invocations of `commit`. I've used `Mockito.clearInvocations` to ensure that the invocations I'm verifying are linked to the correct case; if you know of a better way to do this, please let me know.

cadonna reviewed (commented):
Hi @nicktelford ! Thanks for the updates!

Here my comments!

nicktelford committed:
Add more StreamsConfigTests

nicktelford committed:
Fix formatting of TaskManager

nicktelford committed:
Add TasksTests

nicktelford committed:
Add KeyValueSegmentsTests

nicktelford committed:
Improve StreamThreadTests

nicktelford committed:
Rename TaskManager#needsCommit for clarity

Now called `transactionBuffersExceedCapacity`, to be more illustrative
of the actual functionality.

The boolean argument has been renamed to `predictNextCycle`, to
illustrate that it should only be used once per iteration.

nicktelford committed:
Split TaskManager#transactionBuffersExceedCapacity

No longer uses a boolean argument to "reset prediction".

Now, we cache whether or not the transaction buffers will be exceeded,
and access that cached information directly from the commit() routine.

This ensures that we always update our prediction only once, at the same
point in each iteration.

nicktelford committed:
Add TaskManagerTests

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1836,6 +1842,36 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
         }
     }
 
+    private boolean transactionBuffersExceedCapacity = false;
```
nit: move this with the other fields up at the top

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1836,6 +1842,36 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
+    }
+
+    boolean transactionBuffersWillExceedCapacity() {
+        final boolean transactionBuffersAreUnbounded = maxUncommittedStateBytes < 0;
```
nit: avoid implicit assumptions about the value of sentinel constants. For example what if we need to introduce another sentinel value that means something other than unbounded, and give it a value of -2? Best to define a static constant like `UNBOUNDED_UNCOMMITTED_BYTES = -1` and then check like this:
```suggestion
        final boolean transactionBuffersAreUnbounded = maxUncommittedStateBytes == UNBOUNDED_UNCOMMITTED_BYTES;
```

ableegoldman replied:
And if you define the UNBOUNDED_UNCOMMITTED_BYTES constant in StreamsConfig, perhaps next to the config definition itself, then users can easily discover and use the variable for clear code, rather than hardcoding "-1" and then coming back a year later and wondering what that value means

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1889,7 +1925,8 @@ private int commitTasksAndMaybeUpdateCommittableOffsets(final Collection<Task> t
             return -1;
         } else {
-            return taskExecutor.commitTasksAndMaybeUpdateCommittableOffsets(tasksToCommit, consumedOffsetsAndMetadata);
+            final int committedOffsets = taskExecutor.commitTasksAndMaybeUpdateCommittableOffsets(tasksToCommit, consumedOffsetsAndMetadata);
```
was this change intentional?

nicktelford replied:
An earlier version of this PR reset `lastUncommittedBytes` to `0` between this line and the `return`. I removed that but neglected to simplify it back. Thanks for pointing this out.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1836,6 +1842,36 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
+            return false;
+        }
+
+        // force an early commit if the uncommitted bytes exceeds or is *likely to exceed* the configured threshold
```
I think this method would benefit from slightly more detailed docs/comments. I'm also a bit concerned about over-optimization, but lmk what you think:

First, IIUC, the idea is to use the number of additional uncommitted bytes since the last commit as a heuristic to estimate how many more uncommitted bytes will be added if we don't commit right now, and thus try to commit "just before" we go over the limit. Makes sense on its face, though it's obviously only as good as the buffer size increases are regular/linear. So the question is: will the buffer grow linearly between commits?

Unfortunately I suspect the answer is "no" in many cases -- In fact, the only case in which the buffer will grow by the same amount each iteration is when every insert is a new key. Of course, most KV stores have a bounded keyspace (else they'd grow forever) so we expect to see updates to the same keys. The very first iteration after a commit will show the most growth, and that will taper off as we rebuild the keyset and see fewer and fewer new keys vs in-place updates.

Of course it's not incorrect to commit early, I just wanted to get your thoughts on this. Maybe a good middle ground would be to maintain a running average `deltaBytes`? Or even just cut the previous `deltaBytes` in half? Either way it's still just an approximation, and things like maxUncommittedBytes are always just best-effort in the end.

nicktelford replied:
> First, IIUC, the idea is to use the number of additional uncommitted bytes since the last commit as a heuristic to estimate how many more uncommitted bytes will be added if we don't commit right now, and thus try to commit "just before" we go over the limit.

Correct, this is the intent. One correction though: we're calculating the delta since the last _iteration_ of the StreamThread, not since the last commit.

This behaviour was requested on the mailing list, IIRC by @cadonna, to reduce the risk of OOM between commits. The thinking was that users will  expect `statestore.uncommitted.max.bytes` to be an upper-bound, and likely set it with little headroom. If the commit interval is fairly long (e.g. 30 seconds under ALOS), then there's a good chance we would exceed this limit by a considerable margin.

> [...] so we expect to see updates to the same keys. [...] and see fewer and fewer new keys vs in-place updates.

Writes in RocksDB are the same size, irrespective of whether they are inserts of new keys, or updates to existing keys. The only case I can think of where this may be different is when migrating an existing record from a non-timestamped to a timestamped column family; in which case, the updates would actually be slightly larger, due to the additional tombstone written.

Ultimately, any heuristic we choose will be imperfect. If we track an average, even since the last commit, then it would be less responsive to sudden increases in throughput, potentially allowing an OOM before the average has had a chance to catch up. The current approach attempts to err on the side of caution: we may trigger a commit unnecessarily early on occasion, but we should be less likely to exceed the configured buffer size.

I'm not against changing this heuristic to a running average, or anything else; I'm just not sufficiently convinced it would be better :grin:

ableegoldman replied:
> Writes in RocksDB are the same size, irrespective of whether they are inserts of new keys, or updates to existing keys.

I could be wrong about this, but I thought that writing to an existing key in the active memtable -- ie before it is flushed and/or made immutable  --  is indeed applied as an overwrite, and does not grow the write buffer's overall size? Of course this only applies to patterns in which keys are updated multiple times between flushes, sure. Still, I feel like this is one of the major points of the write buffer: to deduplicate same-key writes as much as possible before flushing them, since once they're flushed it has to wait for a compaction to deduplicate further.

That said, I don't feel strongly about this, and any potential performance degradation should come to light in the benchmarks. Thanks for explaining your thoughts on this

nicktelford replied:
> I could be wrong about this, but I thought that writing to an existing key in the active memtable -- ie before it is flushed and/or made immutable -- is indeed applied as an overwrite, and does not grow the write buffer's overall size? Of course this only applies to patterns in which keys are updated multiple times between flushes, sure. Still, I feel like this is one of the major points of the write buffer: to deduplicate same-key writes as much as possible before flushing them, since once they're flushed it has to wait for a compaction to deduplicate further.

This config controls the maximum memory used by the transaction buffers aka. WriteBatches. These work a bit differently to memtables/write buffers: all writes to a WriteBatch are written serially, even if the same key is touched multiple times. The size of the index it tracks over those writes _may_ vary with cardinality, but it shouldn't be enough to worry about I expect.

It's worth remembering that our tracking of RocksDB transaction buffer size will not be perfect either, because WriteBatch has no means to determine its size in memory. We will be simply tracking the aggregate size of the keys and values being written to it. This means that the index and any serialization overhead used by WriteBatchWithIndex will not be included in this size estimate.

nicktelford committed:
Move TaskManager field to top

nicktelford committed:
Extract unbounded value to constant

nicktelford committed:
Remove unintended refactoring

nicktelford commented:
@cadonna @ableegoldman Is there any more we need to do with this PR before we can merge it?
